### PR TITLE
[bug] fix http-header for application/json

### PR DESCRIPTION
**Pull request by ggjjj**
Bug fix for https://github.com/Azure/iotedge/issues/6505#issuecomment-1184219109

### CODE REVIEW

File: http-common/src/server.rs [hunk 1]
```diff
@@ -113,7 +113,7 @@ macro_rules! make_service {
                                             } else {
                                                 let content_type = headers.get(hyper::header::CONTENT_TYPE).and_then(|value| value.to_str().ok());
 
-                                                match content_type.as_deref() {
+                                                match content_type.as_deref().and_then(|ct| ct.starts_with("application/json").then_some("application/json")) {
```
Comment (arsing) [inline http-common/src/server.rs:116]: This allows any string that starts with `"application/json"`, such as `"application/jsonn"`

**Review by arsing — CHANGES_REQUESTED** [review]

File: http-common/src/server.rs [hunk 2]
```diff
@@ -159,21 +156,18 @@ macro_rules! make_service {
...
-
-
-                                                        let body: <$route as http_common::server::Route>::PostBody = match serde_json::from_slice(&body) {
-                                                            Ok(body) => body,
-                                                            Err(err) => return Ok((http_common::server::Error {
-                                                                status_code: http::StatusCode::UNPROCESSABLE_ENTITY,
-                                                                message: http_common::server::error_to_message(&err).into(),
-                                                            }).to_http_response()),
-                                                        };
+                                                if content_type.as_deref().map_or(true, |content_type| content_type == "application/json" || content_type.starts_with("application/json;")) {
```
Comment (onalante-msft) [inline http-common/src/server.rs:159]:
```suggestion
                                                if content_type.and_then(|ctype| Some(ctype.split(';').next()?.trim() == "application/json")).unwrap_or(true) {
```

**Review by onalante-msft — CHANGES_REQUESTED** [review]

File: http-common/src/server.rs [hunk 3]
```diff
@@ -111,22 +111,19 @@ macro_rules! make_service {
...
-                                                                message: http_common::server::error_to_message(&err).into(),
-                                                            }).to_http_response()),
-                                                        };
-
-                                                        Some(body)
-                                                    },
-                                                    _ => None,
-                                                }
+                                            let content_type = headers.get(hyper::header::CONTENT_TYPE).and_then(|value| value.to_str().ok());
+                                            if content_type.as_deref().map_or(true, |content_type| content_type == "application/json" || content_type.starts_with("application/json;")) {
```
Comment (onalante-msft) [inline http-common/src/server.rs:115]: Same suggestion as other comment.

File: http-common/src/server.rs [hunk 4]
```diff
@@ -185,38 +179,37 @@ macro_rules! make_service {
...
-
-                                                    body
-                                                },
-                                                _ => {
-                                                    return Ok((http_common::server::Error {
-                                                        status_code: http::StatusCode::UNSUPPORTED_MEDIA_TYPE,
-                                                        message: "request body must be application/json".into(),
-                                                    }).to_http_response());
-                                                },
+                                            let body = if content_type.as_deref().map_or(true, |content_type| content_type == "application/json" || content_type.starts_with("application/json;")) {
```
Comment (onalante-msft) [inline http-common/src/server.rs:182]: Same suggestion as other comment.

File: http-common/src/server.rs (hunk 3 above)
Comment (onalante-msft) [inline http-common/src/server.rs:115]: Same suggestion as other comment.

File: http-common/src/server.rs (hunk 4 above)
Comment (onalante-msft) [inline http-common/src/server.rs:182]: Same suggestion as other comment.

Comment (onalante-msft) [inline http-common/src/server.rs:182]:
```suggestion
                                            let body = if content_type.map_or(true, |ctype| ctype.split(';').next().expect("split always returns at least one element").trim() == "application/json") {
```

**Review by onalante-msft — CHANGES_REQUESTED** [review]

**Comment by ggjjj** [comment]
Verified the fix with OWS as well.

**Review by onalante-msft — APPROVED** [review]

**Review by arsing — APPROVED** [review]

**Pull request merged**